src/hotspot/share/opto/vectorization.hpp

```diff
@@ -1315,4 +1320,432 @@ class AlignmentSolver {
+};
+
+struct VTransformBoolTest {
+  const BoolTest::mask _mask;
+  const bool _is_negated;
+  VTransformBoolTest(const BoolTest::mask mask, bool is_negated) :
+    _mask(mask), _is_negated(is_negated) {}
+};
+
+class VTransformBoolVectorNode : public VTransformElementWiseVectorNode {
+private:
+  const VTransformBoolTest _test;
+public:
+  VTransformBoolVectorNode(VTransformGraph& graph, int number_of_nodes, VTransformBoolTest test) :
+    VTransformElementWiseVectorNode(graph, 2, number_of_nodes), _test(test) {}
+  VTransformBoolTest test() const { return _test; }
+  virtual VTransformBoolVectorNode* isa_BoolVector() override { return this; }
+  virtual VTransformApplyStatus apply(const VLoopAnalyzer& vloop_analyzer,
+                                      const GrowableArray<Node*>& vnode_idx_to_transformed_node) const override;
+  NOT_PRODUCT(virtual const char* name() const override { return "BoolVector"; };)
+};
+
+// Reduction of some initial scalar value and a vector, resulting in a new scalar.
+//   val = init;
+//   for (v in vector) { val = val OP v; }
+//   return val;
+class VTransformReductionVectorNode : public VTransformVectorNode {
+public:
+  // req = 3 -> [ctrl, init, vector]
+  VTransformReductionVectorNode(VTransformGraph& graph, int number_of_nodes) :
+    VTransformVectorNode(graph, 3, number_of_nodes) {}
+  virtual VTransformReductionVectorNode* isa_ReductionVector() override { return this; }
+  virtual VTransformApplyStatus apply(const VLoopAnalyzer& vloop_analyzer,
+                                      const GrowableArray<Node*>& vnode_idx_to_transformed_node) const override;
+  NOT_PRODUCT(virtual const char* name() const override { return "ReductionVector"; };)
+};
+
+class VTransformLoadVectorNode : public VTransformVectorNode {
+public:
+  // req = 3 -> [ctrl, mem, adr]
+  VTransformLoadVectorNode(VTransformGraph& graph, int number_of_nodes) :
+    VTransformVectorNode(graph, 3, number_of_nodes) {}
+  LoadNode::ControlDependency control_dependency() const;
+  virtual VTransformApplyStatus apply(const VLoopAnalyzer& vloop_analyzer,
+                                      const GrowableArray<Node*>& vnode_idx_to_transformed_node) const override;
+  NOT_PRODUCT(virtual const char* name() const override { return "LoadVector"; };)
+};
+
+class VTransformStoreVectorNode : public VTransformVectorNode {
+public:
+  // req = 4 -> [ctrl, mem, adr, val]
+  VTransformStoreVectorNode(VTransformGraph& graph, int number_of_nodes) :
+    VTransformVectorNode(graph, 4, number_of_nodes) {}
+  virtual VTransformApplyStatus apply(const VLoopAnalyzer& vloop_analyzer,
+                                      const GrowableArray<Node*>& vnode_idx_to_transformed_node) const override;
+  NOT_PRODUCT(virtual const char* name() const override { return "StoreVector"; };)
+};
+
+// Invoke callback on all memops, in the order of the schedule.
+template<typename Callback>
+void VTransformGraph::for_each_memop_in_schedule(Callback callback) const {
```

Comment:
Note: needed to put this here in the `hpp` so that the `template` could be instanciated, not just for the `vectorization.cpp` but also for `superword.cpp`, where I have still 1 usage. I can probably move around some code in a future RFE to improve this, but for now this reduces the number of lines I need to let you review ;)